daemon/appinfo.c

```diff
@@ -627,51 +646,134 @@ appinfo_clear_permissions(appinfo_t *self)
-    gchar     *path        = NULL;
-    GError    *err         = NULL;
-    gchar    **permissions = NULL;
+static appinfo_file_t
+appinfo_combined_file_state(appinfo_file_t state1, appinfo_file_t state2)
+{
+    /*  state[12] | UNCHANGED |  CHANGED  |  INVALID  |  DELETED  |  MISSING
+     * ===========#===========#===========#===========#===========#==========
+     *  UNCHANGED | UNCHANGED |  CHANGED  |  INVALID  |  CHANGED  | UNCHANGED
+     * ===========#-----------+-----------+-----------+-----------+-----------
+     *   CHANGED  |  CHANGED  |  CHANGED  |  INVALID  |  CHANGED  |  CHANGED
+     * ===========#-----------+-----------+-----------+-----------+-----------
+     *   INVALID  |  INVALID  |  INVALID  |  INVALID  |  INVALID  |  INVALID
+     * ===========#-----------+-----------+-----------+-----------+-----------
+     *   DELETED  |  CHANGED  |  CHANGED  |  INVALID  |  DELETED  |  DELETED
+     * ===========#-----------+-----------+-----------+-----------+-----------
+     *   MISSING  | UNCHANGED |  CHANGED  |  INVALID  |  DELETED  |  MISSING
+     */
 
-    path = path_from_desktop_name(appinfo_id(self));
+    if (state2 <= APPINFO_FILE_MISSING) {
+        switch (state1) {
+        case APPINFO_FILE_UNCHANGED:
+            if( state2 == APPINFO_FILE_DELETED )
+                return APPINFO_FILE_CHANGED;
+            if( state2 == APPINFO_FILE_MISSING )
+                return APPINFO_FILE_UNCHANGED;
+            return state2;
+        case APPINFO_FILE_CHANGED:
+            if( state2 == APPINFO_FILE_INVALID )
+                return APPINFO_FILE_INVALID;
+            return APPINFO_FILE_CHANGED;
+        case APPINFO_FILE_INVALID:
+            return APPINFO_FILE_INVALID;
+        case APPINFO_FILE_DELETED:
+            if( state2 <= APPINFO_FILE_CHANGED )
+                return APPINFO_FILE_CHANGED;
+            if( state2 == APPINFO_FILE_INVALID )
+                return APPINFO_FILE_INVALID;
+            return APPINFO_FILE_DELETED;
+        case APPINFO_FILE_MISSING:
+            return state2;
+        }
+    }
+    log_err("Unknown state: (%d, %d)", state1, state2);
+    return APPINFO_FILE_INVALID;
+}
+
+static appinfo_file_t
+appinfo_check_desktop_from_path(appinfo_t *self, const gchar *path, appinfo_dir_t dir)
+{
+    appinfo_file_t state = APPINFO_FILE_UNCHANGED;
 
     /* Check if the file has changed since last parse */
     struct stat st = {};
     if( stat(path, &st) == -1 ) {
-        log_warning("%s: could not stat: %m", path);
-        if( errno == ENOENT )
-            appinfo_set_state(self, APPINFO_STATE_DELETED);
-        else
-            appinfo_set_state(self, APPINFO_STATE_INVALID);
+        log_debug("%s: could not stat: %m", path);
+        if( errno == ENOENT ) {
+            /* If file existed before, self->anf_dt_ctime[dir] != -1 */
+            state = self->anf_dt_ctime[dir] != -1 ? APPINFO_FILE_DELETED : APPINFO_FILE_MISSING;
+            self->anf_dt_ctime[dir] = -1;
```

Comment:
I agree and will change.

I added this because I needed to know if the file existed before. Removed file could change the data that would be loaded to memory but if file did not exist before it shouldn't make a difference.

Reply:
Added this also after the access() test. Is that okay? I just think that if the file status is deemed invalid, ctime should be set to -1.

Reply:
I guess the original idea with just one file was something like: as long as stat() works, the file is not otherwise re-evaluated until ctime changes. So that in case of access() check leading to invalidation -> file gets locked in to invalid state and the access() check is not repeated until there is a ctime change (due to some change which might make difference for access check). And clearing cached ctime value on access check failure would then mean that there would be a perceived file change requiring full inspection every time. AFAICT the same applies for the two-files case too -> I'd not clear/set the cached ctime except in relation to stat() calls.

Reply:
Right, I did not think that. Thanks! I removed the problematic line.